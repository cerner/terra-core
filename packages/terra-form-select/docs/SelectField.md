# Terra Form Select Field

A convenience wrapper assembling a [terra-form-select](https://github.com/cerner/terra-core/tree/master/packages/terra-form-select) within a [terra-form-field](https://github.com/cerner/terra-core/tree/master/packages/terra-form-field).

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-form-select`

## Component Features

 * [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
 * [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
 * [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
 * [Internationalization Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#internationalization-i18n-support)
 * [Localization Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#internationalization-i18n-support)
 * [LTR/RTL Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#ltr--rtl-support)
