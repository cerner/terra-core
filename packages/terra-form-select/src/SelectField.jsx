import React from 'react';
import PropTypes from 'prop-types';
import Field from 'terra-form-field';
import Select from './Select';
import Variants from './_constants';

const propTypes = {
  /**
   * Whether a clear option is available to clear the selection.
   * This is not applicable to the `multiple` or `tag` variants since the selection can already be deselected using the tag.
   */
  allowClear: PropTypes.bool,
  /**
   * The select options.
   */
  children: PropTypes.node,
  /**
   * The field label.
   */
  label: PropTypes.node.isRequired,
  /**
   * The default value of the select.
   */
  defaultValue: PropTypes.oneOfType([PropTypes.array, PropTypes.number, PropTypes.string]),
  /**
   * Whether the input is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Error message displayed when the select is invalid.
   */
  error: PropTypes.node,
  /**
   * Help message to display with the select.
   */
  help: PropTypes.node,
  /**
   * Whether to hide the required indicator on the label.
   */
  hideRequired: PropTypes.bool,
  /**
   * Whether the field is displayed inline. Displays block by default.
   */
  isInline: PropTypes.bool,
  /**
   * Whether the field is invalid.
   */
  isInvalid: PropTypes.bool,
  /**
   * Whether the label is hidden. Allows hiding the label while meeting accessibility guidelines.
   */
  isLabelHidden: PropTypes.bool,
  /**
   * Additional attributes to spread onto the label.
   */
  // eslint-disable-next-line react/forbid-prop-types
  labelAttrs: PropTypes.object,
  /**
   * The maximum number of options that can be selected. A value less than 2 will be ignored.
   * Only applicable to variants allowing multiple selections (e.g.; `multiple`; `tag`).
   */
  maxSelectionCount: PropTypes.number,
  /**
   * Set the max-width of a field using `length` or `%`.  Best practice recommendation to never exceed
   * a rendered value of 1020px. _(Note: Providing custom inline styles will take precedence.)_
   */
  maxWidth: PropTypes.string,
  /**
   * Callback function triggered when the select value changes. function(value)
   */
  onChange: PropTypes.func,
  /**
   * Placeholder text.
   */
  placeholder: PropTypes.string,
  /**
   * Whether the field is required.
   */
  required: PropTypes.bool,
  /**
   * Additional attributes to spread onto the select.
   */
  // eslint-disable-next-line react/forbid-prop-types
  selectAttrs: PropTypes.object,
  /**
   * The Select identifier. Links the htmlFor of the field to the select identifier.
   */
  selectId: PropTypes.string.isRequired,
  /**
   * Whether to append the 'optional' label to a non-required field label.
   */
  showOptional: PropTypes.bool,
  /**
   * The value of the select.
   */
  value: PropTypes.oneOfType([PropTypes.array, PropTypes.number, PropTypes.string]),
  /**
   * The select variant. One of `default`, `combobox`, `multiple`, `tag`, or `search`.
   */
  variant: PropTypes.oneOf([
    Variants.COMBOBOX,
    Variants.DEFAULT,
    Variants.MULTIPLE,
    Variants.SEARCH,
    Variants.TAG,
  ]),
};

const defaultProps = {
  allowClear: false,
  children: undefined,
  defaultValue: undefined,
  disabled: false,
  error: undefined,
  help: undefined,
  hideRequired: false,
  isInline: false,
  isInvalid: false,
  isLabelHidden: false,
  labelAttrs: {},
  maxSelectionCount: undefined,
  maxWidth: undefined,
  onChange: undefined,
  placeholder: undefined,
  required: false,
  selectAttrs: {},
  showOptional: false,
  value: undefined,
  variant: 'default',
};

const contextTypes = {
  /* eslint-disable consistent-return */
  intl: (context) => {
    if (context.intl === undefined) {
      return new Error('Component is internationalized, and must be wrapped in terra-base');
    }
  },
};


const SelectField = ({
  allowClear,
  children,
  defaultValue,
  disabled,
  error,
  help,
  hideRequired,
  isInline,
  isInvalid,
  isLabelHidden,
  label,
  labelAttrs,
  maxSelectionCount,
  maxWidth,
  onChange,
  placeholder,
  required,
  selectAttrs,
  selectId,
  showOptional,
  value,
  variant,
  ...customProps
}, context) => {
  let helpText = help;
  if (maxSelectionCount !== undefined && maxSelectionCount >= 2) {
    const limitSelectionText = context.intl.formatMessage({ id: 'Terra.form.select.maxSelectionHelp' }, { text: maxSelectionCount });

    if (help) {
      helpText = (
        <span>
          { limitSelectionText }
          { ' ' }
          { help }
        </span>
      );
    } else {
      helpText = limitSelectionText;
    }
  }

  let ariaDescriptionIds;

  if (help && error && isInvalid) {
    ariaDescriptionIds = `${selectId}-error ${selectId}-help`;
  } else {
    if (help) {
      ariaDescriptionIds = `${selectId}-help`;
    }

    if (error && isInvalid) {
      ariaDescriptionIds = `${selectId}-error`;
    }
  }

  return (
    <Field
      {...customProps}
      label={label}
      labelAttrs={labelAttrs}
      error={error}
      help={helpText}
      hideRequired={hideRequired}
      required={required}
      showOptional={showOptional}
      isInvalid={isInvalid}
      isInline={isInline}
      isLabelHidden={isLabelHidden}
      htmlFor={selectId}
      maxWidth={maxWidth}
    >
      <Select
        {...selectAttrs}
        ariaLabel={label}
        allowClear={allowClear}
        aria-describedby={ariaDescriptionIds}
        disabled={selectAttrs.disabled || disabled}
        id={selectId}
        isInvalid={isInvalid}
        defaultValue={defaultValue}
        maxSelectionCount={maxSelectionCount !== undefined && maxSelectionCount < 2 ? undefined : maxSelectionCount}
        onChange={onChange}
        placeholder={placeholder}
        value={value}
        variant={variant}
      >
        {children}
      </Select>
    </Field>
  );
};

SelectField.propTypes = propTypes;
SelectField.defaultProps = defaultProps;
SelectField.contextTypes = contextTypes;

SelectField.Option = Select.Option;
SelectField.OptGroup = Select.OptGroup;

export default SelectField;
