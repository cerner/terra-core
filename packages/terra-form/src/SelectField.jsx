import React from 'react';
import PropTypes from 'prop-types';

import Field from './Field';
import Select from './Select';

const propTypes = {

  /**
   * List of choices to be selected.
   * If choices and options array are supplied, options array will be used over choices.
   */
  choices: PropTypes.array,

  /**
   * List of object key and value pairs for choices to be selected.
   * If choices and options array are supplied, options array will be used over choices.
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    display: PropTypes.string.isRequired,
  })),

  /**
   * Error message for when the input is invalid.
   */
  error: PropTypes.node,

  /**
   *  Help element to display with the input.
   */
  help: PropTypes.node,

  /**
   * Whether the field is inline.
   */
  isInline: PropTypes.bool,

  /**
   * Label of the input.
   */
  label: PropTypes.node,

  /**
   * Attributes to attach to the label.
   */
  labelAttrs: PropTypes.object,

  /**
   * Function to trigger when the user changes the select value. Provide a function to create a controlled input.
   */
  onChange: PropTypes.func,

  /**
   * Name of the select field.
   */
  name: PropTypes.string,

  /**
   * Custom attributes to apply to the select.
   */
  selectAttrs: PropTypes.object,

  /**
   * Whether the select is required or not.
   */
  required: PropTypes.bool,

  /**
   * The value to start the select on.
   */
  defaultValue: PropTypes.string,

  /**
   * The value of the select element. Use this to create a controlled input.
   */
  value: PropTypes.string,
};

const defaultProps = {
  choices: null,
  options: null,
  error: null,
  help: null,
  isInline: false,
  label: null,
  labelAttrs: {},
  onChange: undefined,
  name: null,
  required: false,
  selectAttrs: {},
  defaultValue: undefined,
  value: undefined,
};

const SelectField = ({
  choices,
  options,
  error,
  help,
  isInline,
  label,
  labelAttrs,
  onChange,
  name,
  required,
  selectAttrs,
  defaultValue,
  value,
  ...customProps
}) => (
  <Field
    {...customProps}
    error={error}
    help={help}
    isInline={isInline}
    label={label}
    labelAttrs={labelAttrs}
    required={required}
  >
    <Select
      {...selectAttrs}
      choices={choices}
      options={options}
      onChange={onChange}
      name={name}
      required={required}
      defaultValue={defaultValue}
      value={value}
    />
  </Field>
);


SelectField.propTypes = propTypes;
SelectField.defaultProps = defaultProps;

export default SelectField;
