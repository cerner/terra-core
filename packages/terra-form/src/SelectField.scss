@import '~terra-mixins';

.terra-Form-select {
  background-color: $terra-white;
  border: $terra-grey-15 solid 1px;
  border-radius: 0.25em;
  font-size: 1em;
  line-height: 1.428em;
  padding: 0.178em;
  width: 100%;
}
