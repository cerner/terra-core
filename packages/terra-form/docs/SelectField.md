# Select Field

Select Field component for rendering selects with standard structured style. Uses the select component. Also exposes attributes specific to the select.

## Usage

```jsx
import React from 'react';
import SelectField from 'terra-form/lib/SelectField';

<SelectField
  options={[{ value: 'm', display: 'moo' },
            { value: 'b', display: 'boo' },
            { value: 'z', display: 'zar' }]}
  error="This field is required"
  name="description"
  label="Description"
  selectAttrs={{ className: 'healtheintent-application' }}
  defaultValue="b"
  required
/>
```
