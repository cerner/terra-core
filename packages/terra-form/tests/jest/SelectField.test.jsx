import React from 'react';
import SelectField from '../../src/SelectField';

it('should render a default SelectField component', () => {
  const selectField = <SelectField options={[{ value: 'm', display: 'moo' }]} />;
  const wrapper = shallow(selectField);
  expect(wrapper).toMatchSnapshot();
});

it('should render a SelectField when all the possible props are passed into it', () => {
  const selectField = (
    <SelectField
      options={[{ value: 'm', display: 'moo' },
                { value: 'b', display: 'boo' }]}
      error="jinkies!"
      isInline
      label="Meddling Kids"
      labelAttrs={{ className: 'mystery-van' }}
      help="solve the mystery"
      name="zounds"
      required
      selectAttrs={{ className: 'scooby-snacks' }}
    />);

  const wrapper = shallow(selectField);
  expect(wrapper).toMatchSnapshot();
});

it('should render as uncontrolled when just the required fields and a default value is passed into the SelectField', () => {
  const selectField = <SelectField options={[{ value: 'm', display: 'moo' }, { value: 'b', display: 'boo' }]} defaultValue="b" />;
  const wrapper = mount(selectField);
  expect(wrapper).toMatchSnapshot();
});

it('should render as controlled when just the required fields and a onChange function is passed into the SelectField', () => {
  const selectField = <SelectField options={[{ value: 'm', display: 'moo' }, { value: 'b', display: 'boo' }]} onChange={() => {}} />;
  const wrapper = mount(selectField);
  expect(wrapper).toMatchSnapshot();
});
