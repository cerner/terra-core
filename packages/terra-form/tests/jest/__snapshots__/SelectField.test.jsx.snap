exports[`test should render a SelectField when all the possible props are passed into it 1`] = `
<Field
  error="jinkies!"
  help="solve the mystery"
  isInline={true}
  label="Meddling Kids"
  labelAttrs={
    Object {
      "className": "mystery-van",
    }
  }
  required={true}>
  <Select
    choices={null}
    className="scooby-snacks"
    name="zounds"
    options={
      Array [
        Object {
          "display": "moo",
          "value": "m",
        },
        Object {
          "display": "boo",
          "value": "b",
        },
      ]
    }
    required={true} />
</Field>
`;

exports[`test should render a default SelectField component 1`] = `
<Field
  error={null}
  help={null}
  isInline={false}
  label={null}
  labelAttrs={Object {}}
  required={false}>
  <Select
    choices={null}
    name={null}
    options={
      Array [
        Object {
          "display": "moo",
          "value": "m",
        },
      ]
    }
    required={false} />
</Field>
`;

exports[`test should render as controlled when just the required fields and a onChange function is passed into the SelectField 1`] = `
<SelectField
  choices={null}
  error={null}
  help={null}
  isInline={false}
  label={null}
  labelAttrs={Object {}}
  name={null}
  onChange={[Function]}
  options={
    Array [
      Object {
        "display": "moo",
        "value": "m",
      },
      Object {
        "display": "boo",
        "value": "b",
      },
    ]
  }
  required={false}
  selectAttrs={Object {}}>
  <Field
    error={null}
    help={null}
    isInline={false}
    label={null}
    labelAttrs={Object {}}
    required={false}>
    <div
      className="field">
      <Select
        choices={null}
        name={null}
        onChange={[Function]}
        options={
          Array [
            Object {
              "display": "moo",
              "value": "m",
            },
            Object {
              "display": "boo",
              "value": "b",
            },
          ]
        }
        required={false}>
        <select
          className="select"
          name={null}
          onChange={[Function]}
          required={false}>
          <option
            value="m">
            moo
          </option>
          <option
            value="b">
            boo
          </option>
        </select>
      </Select>
    </div>
  </Field>
</SelectField>
`;

exports[`test should render as uncontrolled when just the required fields and a default value is passed into the SelectField 1`] = `
<SelectField
  choices={null}
  defaultValue="b"
  error={null}
  help={null}
  isInline={false}
  label={null}
  labelAttrs={Object {}}
  name={null}
  options={
    Array [
      Object {
        "display": "moo",
        "value": "m",
      },
      Object {
        "display": "boo",
        "value": "b",
      },
    ]
  }
  required={false}
  selectAttrs={Object {}}>
  <Field
    error={null}
    help={null}
    isInline={false}
    label={null}
    labelAttrs={Object {}}
    required={false}>
    <div
      className="field">
      <Select
        choices={null}
        defaultValue="b"
        name={null}
        options={
          Array [
            Object {
              "display": "moo",
              "value": "m",
            },
            Object {
              "display": "boo",
              "value": "b",
            },
          ]
        }
        required={false}>
        <select
          className="select"
          defaultValue="b"
          name={null}
          required={false}>
          <option
            value="m">
            moo
          </option>
          <option
            value="b">
            boo
          </option>
        </select>
      </Select>
    </div>
  </Field>
</SelectField>
`;
